Support asterisk in principals for trust domain migration

pitlv2109 opened the pull request:
Support asterisk in principals for trust domain migration. A principal can be `*, */ns/foo/sa/bar, or */bar`, etc. 

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | a20fb3b6dc1930e94dcc8e922277c0b78d1df38f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/2600) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | a20fb3b6dc1930e94dcc8e922277c0b78d1df38f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/2600) | `/test unit-tests_istio`
e2e-bookInfoTests-envoyv2-v1alpha3_istio | a20fb3b6dc1930e94dcc8e922277c0b78d1df38f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/e2e-bookInfoTests-envoyv2-v1alpha3_istio/2491) | `/test e2e-bookInfoTests-envoyv2-v1alpha3_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented on pilot/pkg/security/authz/builder/testdata/v1beta1/simple-policy-principal-with-*-in.yaml:
```diff
@@ -0,0 +1,16 @@
+        - source:
+            principals: ["*"]
+        - source:
+            principals: ["*/ns/foo/sa/rule[0]/sa/rule[0]-from[1]-principal[0]", "*bar/ns/foo/sa/rule[0]-from[1]-principal[1]"]
```
replace `*/ns/foo/sa/rule[0]/sa/rule[0]-from[1]-principal[0]` with `*/ns/foo/sa/rule[0]-from[1]-principal[0]`

yangminzhu commented on pilot/pkg/security/authz/builder/testdata/v1beta1/simple-policy-principal-with-*-out.yaml:
```diff
@@ -0,0 +1,31 @@
+                          - key: source.principal
+                        value:
+                          stringMatch:
+                            suffix: /ns/foo/sa/rule[0]/sa/rule[0]-from[1]-principal[0]
```
Same here.

yangminzhu commented on pilot/pkg/security/authz/builder/testdata/v1beta1/simple-policy-principal-with-*-out.yaml:
```diff
@@ -0,0 +1,31 @@
+                        value:
+                          stringMatch:
+                            suffix: /ns/foo/sa/rule[0]/sa/rule[0]-from[1]-principal[0]
+                    - metadata:
```
For `*bar/ns/foo/sa/rule[0]-from[1]-principal[1]`, there is only 1 principal `suffix: bar/ns/foo/sa/rule[0]-from[1]-principal[1] `, this seems different from the case in `simple-policy-user-with-*-out.yaml`?

yangminzhu commented on pilot/pkg/security/authz/model/model.go:
```diff
@@ -287,14 +287,21 @@ func replaceTrustDomainAliases(trustDomain string, trustDomainAliases []string,
 		trustDomainFromPrincipal := getTrustDomain(principal)
-		//TODO(pitlv2109): Handle * and prefix/suffix.
-		if trustDomainFromPrincipal == "" {
+		// Return the existing principals if getTrustDomain() returns an empty string or "*".
```
Before adding more changes, could you first move replaceTrustDomainAliases() and related functions out of the model.go and put them either in util.go or in a different file?

pitlv2109 replied:
I have the change locally. I'll send a PR out and we can get that in first.

pitlv2109 replied:
I sent out a PR here https://github.com/istio/istio/pull/18066. Thank you!

yangminzhu commented on pilot/pkg/security/authz/model/model.go:
```diff
@@ -308,7 +315,15 @@ func replaceTrustDomainAliases(trustDomain string, trustDomainAliases []string,
 	principalsForAliases := []string{}
 	for _, tdAlias := range trustDomainAliases {
-		principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(tdAlias, principal))
+		tdWithAsteriskPrefix := getTrustDomain(principal)
```
This could be done out of the loop.

pitlv2109 replied:
Thanks

yangminzhu commented on pilot/pkg/security/authz/model/util.go:
```diff
@@ -27,14 +27,14 @@ import (
 }
 
-// prefixMatch checks if string "a" prefix matches "pattern".
+// prefixMatch checks if string "a" suffix matches "pattern".
```
Why change the comment?

pitlv2109 replied:
Looks like you have a typo before. Both prefixMatch and suffixMatch have the same comments https://github.com/istio/istio/blob/master/pilot/pkg/security/authz/model/util.go#L46. `prefixMatch` checks if foobar matches the pattern *bar.

yangminzhu replied:
But this should be changed in the `suffixMatch` function but not here.

Also better to rewrite the comment to: `prefixMatch checks if pattern is a prefix match and if string a has the given prefix`, same for `suffixMatch`.

yangminzhu commented on pilot/pkg/security/authz/model/util.go:
```diff
@@ -27,14 +27,14 @@ import (
 func stringMatch(a string, list []string) bool {
 	for _, s := range list {
-		if a == s || s == "*" || prefixMatch(a, s) || suffixMatch(a, s) {
+		if a == s || s == "*" || prefixMatch(a, s) || prefixMatch(s, a) || suffixMatch(a, s) || suffixMatch(s, a) {
```
Why change this?

pitlv2109 replied:
If you look at the tests, you can see that previously the pattern can only exist in the list. I extended to allow pattern can exist in `a` too. So now it supports both cases:
`S: "*local", List: []string{"cluster.local"}` and `S: "cluster.local", List: []string{"*local"}`. I can write another function to extend this and use it, or use the existing one.

yangminzhu replied:
I know what the changes you made here I'm asking why make the change in this PR?

The change will affect other place: https://github.com/istio/istio/blob/96109c83af15b9c431aef6428127e606403c9d57/pilot/pkg/security/authz/model/permission.go#L52 so it's changing the API of the policy.

pitlv2109 replied:
I made the change because I thought it wouldn't affect the current implementation, because `service.Name` cannot have prefix/suffix if I understand it correctly? But I'll create a different function for this PR.

yangminzhu replied:
I suggest not to change this and depend on external behavior. You could create a separate function or even just embed it in your main function as it's only used by your code.

pitlv2109 replied:
Will do. Thanks!

yangminzhu commented on pilot/pkg/security/authz/policy/v1alpha1/v1alpha1_test.go:
```diff
@@ -209,6 +209,35 @@ func TestV1alpha1Generator_Generate(t *testing.T) {
+			trustDomain:        "td1",
+			trustDomainAliases: []string{"cluster.local", "td2"},
+			wantRules: map[string][]string{
+				"role": {policy.RoleTag("role"),
```
The format is strange, could you add a newline?

yangminzhu commented on pilot/pkg/security/authz/policy/v1alpha1/v1alpha1_test.go:
```diff
@@ -209,6 +209,35 @@ func TestV1alpha1Generator_Generate(t *testing.T) {
+			trustDomainAliases: []string{"cluster.local", "td2"},
+			wantRules: map[string][]string{
+				"role": {policy.RoleTag("role"),
+					"*"},
```
Please do not just put a raw string here, the test `policy.Verify()` is doing string search to find some specific pattern in the result, that's why we need those `policy.RoleTag` to make it specific, otherwise it could lead to false positive results in the test that the search matched with pattern that is not from your input.

yangminzhu commented on pilot/pkg/security/authz/policy/v1alpha1/v1alpha1_test.go:
```diff
@@ -47,21 +47,21 @@ func TestV1alpha1Generator_Generate(t *testing.T) {
 			policies: []*model.Config{
 				policy.SimpleRole("role", namespaceB, serviceFoo),
-				policy.SimpleBinding("binding", namespaceB, "role"),
+				policy.SimpleBinding("binding", namespaceB, "role", policy.SimplePrincipal("binding")),
```
It's really a pain looking at these extra call to `policy.SimplePrincipal`, especially it's used in 90% cases, could you refactor to avoid these duplicate calling?

liminw commented on pilot/pkg/security/authz/model/model.go:
```diff
@@ -308,7 +315,15 @@ func replaceTrustDomainAliases(trustDomain string, trustDomainAliases []string,
+		// with the prefix * trust domain, rather than the trust domain from the aliases list.
+		// The aliases list should contain fully qualified trust domain names (no prefix or suffix).
+		if suffixMatch(tdAlias, tdWithAsteriskPrefix) {
+			principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(tdWithAsteriskPrefix, principal))
```
This basically says, if tdWithAsteriskPrefix matches any of the tdAlias (including trustDomain), add the original principal (including the *), in addition, add modified principal with tdAliases. 

replaceTrustDomainInPrincipal(tdWithAsteriskPrefix, principal) is a no-op.

pitlv2109 commented on pilot/pkg/security/authz/model/model.go:
```diff
@@ -287,14 +287,21 @@ func replaceTrustDomainAliases(trustDomain string, trustDomainAliases []string,
 		// or if the extracted/existing trust domain is "cluster.local", which is a pointer to the local trust domain
 		// and its aliases.
-		if found(trustDomainFromPrincipal, trustDomainAliases) || trustDomainFromPrincipal == "cluster.local" {
+		if stringMatch(trustDomainFromPrincipal, trustDomainAliases) || trustDomainFromPrincipal == "cluster.local" {
```
1. Why does the trust domain from the existing policy needs to match against the new trust domain? It *should* match one of the aliases, and we generate new policy for the new trust domain, right? 

2. If we have a principal "\*/ns/foo/sa/bar", and an alias is "old-td", it's a match, right?

liminw replied:
1. You will not be able to predict when the policy is written. It is possible that the users write policies targeting for the new trust domain. The principle is, trustDomain and trustDomainAliases should be treated the same, when we convert the policies. 

2. If principal is "*/ns/foo/sa/bar", you do not need to expand the policy, it already matches any trust domain. 
If principal is "*td/ns/foo/sa/bar", and aliases is ["old-td", "new"], you should expand the principal to include both "new/ns/foo/sa/bar", in addition to "*td/ns/foo/sa/bar". 

So the logic should be:
Let's call (trustDomain + trustDomainAliases) "tdList"
If trustDomainFromPrincipal matches any of tdList, principalList = original principal + any principals constructed from the tdList (and not included in the original principal).
If trustDomainFromPrincipal does not match any of tdList, principalList = original principal.

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio_e2e_cloudfoundry_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/istio_e2e_cloudfoundry_istio/2976) | `/test istio_e2e_cloudfoundry_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio_e2e_cloudfoundry_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/istio_e2e_cloudfoundry_istio/2976) | `/test istio_e2e_cloudfoundry_istio`
unit-tests_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/3173) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
@yangminzhu  @liminw I pushed some code to address your comments. In particular, I moved all trust domain alias stuff out of existing unit tests and added new tests instead. Also implemented some extra requirements for * prefix/suffix as we discussed offline yesterday.

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio_e2e_cloudfoundry_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/istio_e2e_cloudfoundry_istio/2976) | `/test istio_e2e_cloudfoundry_istio`
unit-tests_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/3173) | `/test unit-tests_istio`
release-test_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/release-test_istio/2993) | `/test release-test_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
istio_e2e_cloudfoundry_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/istio_e2e_cloudfoundry_istio/2976) | `/test istio_e2e_cloudfoundry_istio`
unit-tests_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/3173) | `/test unit-tests_istio`
release-test_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/release-test_istio/2993) | `/test release-test_istio`
pilot-e2e-envoyv2-v1alpha3_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/pilot-e2e-envoyv2-v1alpha3_istio/3028) | `/test pilot-e2e-envoyv2-v1alpha3_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-test_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/release-test_istio/2993) | `/test release-test_istio`
pilot-e2e-envoyv2-v1alpha3_istio | 03985c33c6e1f40a026c1ddee9d20cc4044fe23a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/pilot-e2e-envoyv2-v1alpha3_istio/3028) | `/test pilot-e2e-envoyv2-v1alpha3_istio`
integ-istioio-k8s-tests_istio | 21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/integ-istioio-k8s-tests_istio/731) | `/test integ-istioio-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-istioio-k8s-tests_istio | 21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/integ-istioio-k8s-tests_istio/731) | `/test integ-istioio-k8s-tests_istio`
integ-new-install-k8s-tests_istio | 21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/integ-new-install-k8s-tests_istio/2357) | `/test integ-new-install-k8s-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

yangminzhu commented on pilot/pkg/security/trustdomain/bundle.go:
```diff
@@ -115,17 +134,17 @@ func getTrustDomain(principal string) string {
+// SPIFFE format and the trust domain cannot be * (i.g. */ns/foo/sa/bar).
+func getTrustDomainOrError(principal string) (string, error) {
+	trustDomain := getTrustDomain(principal)
+	if trustDomain == "" || trustDomain == "*" {
```
`*` is considered invalid but something like `*-suffix` is still valid?

yangminzhu commented on pilot/pkg/security/trustdomain/bundle_test.go:
```diff
@@ -32,6 +32,18 @@ func TestReplaceTrustDomainAliases(t *testing.T) {
+			expect:            []string{"*"},
+		},
+		{
+			name:              "Principal with * prefix and right format",
```
just `Principal with prefix` is okay.

yangminzhu commented on pilot/pkg/security/trustdomain/bundle_test.go:
```diff
@@ -44,13 +56,25 @@ func TestReplaceTrustDomainAliases(t *testing.T) {
+			name:              "Two trust domain aliases with * prefix in trust domain",
+			trustDomainBundle: NewTrustDomainBundle("td2", []string{"foo-td1", "cluster.local"}),
+			principals:        []string{"*-td1/ns/foo/sa/bar"},
+			expect:            []string{"td2/ns/foo/sa/bar", "*-td1/ns/foo/sa/bar", "cluster.local/ns/foo/sa/bar"},
```
Say we have td = `td2` and alias = `["foo-td1", "cluster.local"]`
- for principal `*-td1/ns/foo/sa/bar` we will generate the output with both "foo-td1" and "cluster.local", even "cluster.local" is not matched
- for principal `td2/ns/foo/sa/bar`. it's still not matched but we will only generate the output with `td2/ns/foo/sa/bar` which seems a little inconsistent from the first case

I'm not following the trust domain design but just feel a little bit strange here, do you know what's the rational for the inconsistent?

pitlv2109 replied:
The logic here is this: for principal `*-td1/ns/foo/sa/bar`, it matches against the new trust domain or one of the aliases (in this case foo-td1). `td2`, `foo-td1`, and `cluster.local` are treated the same, hence the generated principals: "td2/ns/foo/sa/bar", "*-td1/ns/foo/sa/bar", "cluster.local/ns/foo/sa/bar"

@liminw

liminw replied:
Yes. this is correct.

yangminzhu commented on pilot/pkg/security/trustdomain/util.go:
```diff
@@ -0,0 +1,50 @@
+// 3. Prefix match. For example, "book*" matches "bookstore", "bookshop", etc.
+// 4. Suffix match. For example, "*/review" matches "/bookstore/review", "/products/review", etc.
+// This is an extensive version of model.stringMatch(). The pattern can be in the string or the list.
+func stringMatch(a string, list []string) bool {
```
This has a different behavior than the `model.stringMatch()`, we should also give it a different name to avoid confusion, also for other functions in this file.

pitlv2109 replied:
It's only used inside this package, so I think it's fine this way (also the comment mentions the stringMatch() from the model package)? If not, what would you name it?

yangminzhu commented on pilot/pkg/security/authz/policy/helper.go:
```diff
@@ -194,7 +199,7 @@ func SimpleAuthzPolicy(name string, namespace string) *model.Config {
 }
 
-func Verify(got *envoy_rbac.RBAC, want map[string][]string) error {
+func Verify(got *envoy_rbac.RBAC, want map[string][]string, arePrincipalsImportant bool) error {
```
the `arePrincipalsImportant` reads quite strange (by reading the name I have no idea what it is going to do), can you use more descriptive name?

pitlv2109 replied:
This mean if we care about the true principal values (which is important for trust domain migration). What would you name it?

yangminzhu commented on pilot/pkg/security/authz/policy/helper.go:
```diff
@@ -229,3 +237,32 @@ func Verify(got *envoy_rbac.RBAC, want map[string][]string) error {
+// in v1beta1, principals can be a list.
+func checkPrincipalCounts(ruleKey string, actual string, want []string) error {
+	// Only work with principals, not other identities like namespaces.
+	keyRegEx := regexp.MustCompile("source.principal")
```
This is hard to maintain and could easily cause false positive/negative, can we actually check the appearance of the principal in `want`?

pitlv2109 replied:
Currently, `want` outputs something like this `[MethodFromRole[role] -td/ns/foo/sa/bar td1/ns/foo/sa/bar td2/ns/foo/sa/bar]`. Would it be better if we add some prefix tag before the principal, like `Principal[-td/ns/foo/sa/bar]`? So that we can easily extract the principals out?

pitlv2109 replied:
Why is it hard to maintain? Previously we were checking the principal from `want`, but something `actualStr` may have more principals than `want`, which is a false positive. For example. want = [foo], actualStr = "....source.principal: foo....source.principal: bar".

yangminzhu commented on pilot/pkg/security/authz/policy/helper.go:
```diff
@@ -124,15 +125,19 @@ func SimpleRole(name string, namespace string, service string) *model.Config {
+	return fmt.Sprintf("UserFromBinding[%s]", name)
+}
+
+func SimpleBinding(name, namespace, role, user string) *model.Config {
```
Can you not to change this function? you can create a `SimpleBindingWithUser()`, and change `SimpleBinding()` to call it, then you don't need the special check of `""` and also do not need to pass the extra `""` in the tests.

pitlv2109 replied:
Sure

yangminzhu commented on pilot/pkg/security/trustdomain/bundle.go:
```diff
@@ -26,21 +26,22 @@ var (
 	return Bundle{
-		TrustDomain:        trustDomain,
-		TrustDomainAliases: trustDomainAliases,
+		// Put the new trust domain to the beginning of the list to avoid changing existing tests.
```
Is this only for test or is it a required property for this data structure?

pitlv2109 replied:
Mostly just for tests. Most tests are written with the new trust domain to appear first.

liminw commented on pilot/pkg/security/trustdomain/bundle.go:
```diff
@@ -115,17 +134,17 @@ func getTrustDomain(principal string) string {
-			return true
-		}
+// getTrustDomainOrError is similar to getTrustDomain. However, it expects principal to have the right
+// SPIFFE format and the trust domain cannot be * (i.g. */ns/foo/sa/bar).
```
Why trust domain cannot be *? "*/ns/foo/sa/bar" looks valid.

pitlv2109 replied:
For the given principal, the trust domain is "", and we already check for that case here https://github.com/istio/istio/blob/21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7/pilot/pkg/security/trustdomain/bundle.go#L65.

This function `getTrustDomainOrError` simply finds trust domain that are not "*" or "" (these trust domains mean they will match to any trust domain).

pitlv2109 replied:
Actually, "/ns/foo/sa/bar" is different from "*/ns/foo/sa/bar". The first one is exact match, the second one is suffix match. The current PR doesn't handle this. I'll push some code for this.

pitlv2109 replied:
Pushed. PTAL.

liminw replied:
Actually I was asking about "\*", and "\*/ns/foo/sa/bar". I forgot to put the escape character.

liminw commented on pilot/pkg/security/trustdomain/bundle.go:
```diff
@@ -78,11 +83,25 @@ func (t Bundle) ReplaceTrustDomainAliases(principals []string) []string {
+			principalsForAliases = append(principalsForAliases, principal)
+		} else {
+			principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(td, principal))
+		}
```
Let's move the logic of appending the original principal outside of the loop. 
* The original principal should always be added. 
* If tdFromPrincipal does not match any TD, simply keep the original principal. 
* If tdFromPrincipal matches any TD, add those that are not included in the original principal. 

We should have corresponding test cases for the above too. 

```
principalsForAliases = append(principalsForAliases, principal)
foundMatch := false
for _, td := range t.TrustDomains {
   if suffixMatch(td, trustDomainFromPrincipal) {
       foundMatch = true
   } 
}
if foundMatch {
   for _, td := range t.TrustDomains {
      if !suffixMatch(td, trustDomainFromPrincipal) {
          principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(td, principal))
      }
   }
}
```

liminw replied:
Do we have a separate function for exact match?

pitlv2109 replied:
The code already supports the 3 cases you mentioned and has tests here https://github.com/istio/istio/blob/21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7/pilot/pkg/security/trustdomain/bundle_test.go#L79 (I should change it to `principals not match any trust domain` instead).

> Do we have a separate function for exact match?

Yes, the code supports exact match, prefix match, and suffix match.

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-istioio-k8s-tests_istio | 21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/integ-istioio-k8s-tests_istio/731) | `/test integ-istioio-k8s-tests_istio`
integ-new-install-k8s-tests_istio | 21ddfc6d4744b0a08dbd2fa628cf3c30ce5904d7 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/integ-new-install-k8s-tests_istio/2357) | `/test integ-new-install-k8s-tests_istio`
pilot-multicluster-e2e_istio | 871568b66791b2f6c60027c25d16ad4444011159 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/pilot-multicluster-e2e_istio/2444) | `/test pilot-multicluster-e2e_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
/test pilot-multicluster-e2e_istio

pitlv2109 commented:
@liminw @yangminzhu I pushed some code to address some comments, also left some comments on your questions. Let me know if you guys have more thoughts on this. Thank you!

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | b201b029025171dbc9a6a196f9623a76e7119df6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/3370) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | b201b029025171dbc9a6a196f9623a76e7119df6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/3370) | `/test unit-tests_istio`
pilot-e2e-envoyv2-v1alpha3_istio | b201b029025171dbc9a6a196f9623a76e7119df6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/pilot-e2e-envoyv2-v1alpha3_istio/3216) | `/test pilot-e2e-envoyv2-v1alpha3_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
/test unit-tests_istio
/test pilot-e2e-envoyv2-v1alpha3_istio

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pilot-e2e-envoyv2-v1alpha3_istio | b201b029025171dbc9a6a196f9623a76e7119df6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/pilot-e2e-envoyv2-v1alpha3_istio/3216) | `/test pilot-e2e-envoyv2-v1alpha3_istio`
unit-tests_istio | b201b029025171dbc9a6a196f9623a76e7119df6 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/3378) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
/test unit-tests_istio

liminw commented on pilot/pkg/security/authz/policy/helper.go:
```diff
@@ -216,8 +238,11 @@ func Verify(got *envoy_rbac.RBAC, want map[string][]string) error {
+				// Enable principal count check if principals are important for the current test (e.g. generating RBAC config based
+				// on trust domain aliases).
+				if arePrincipalsImportant {
+					err = checkPrincipalCounts(key, actualStr, values)
```
Did you check the expected principals already? Why do you need to check principal count?

pitlv2109 replied:
This comment explains this in more detail https://github.com/istio/istio/pull/18055#discussion_r337757306

pitlv2109 replied:
@liminw I added a commit to change the name to `needToCheckPrincipals` instead, also another commit to reflect the presence match PR that got merged recently. PTAL and let me know if you have any more comments. Thanks!

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | f495027f55d9e99ad3c63e62984b6b6be5e9535f | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18011/unit-tests_istio/3436) | `/test unit-tests_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
/test unit-tests_istio

istio-testing commented:
In response to a cherrypick label: #18011 failed to apply on top of branch "release-1.4":
```Applying: Handle * in principals
Applying: Add tests
.git/rebase-apply/patch:449: new blank line at EOF.
+
warning: 1 line adds whitespace errors.
error: Failed to merge in the changes.
Using index info to reconstruct a base tree...
M	pilot/pkg/security/authz/builder/builder_test.go
M	pilot/pkg/security/authz/builder/integration_test.go
M	pilot/pkg/security/authz/model/model.go
M	pilot/pkg/security/authz/model/util.go
M	pilot/pkg/security/authz/model/util_test.go
M	pilot/pkg/security/authz/policy/v1alpha1/v1alpha1_test.go
Falling back to patching base and 3-way merge...
Auto-merging pilot/pkg/security/authz/policy/v1alpha1/v1alpha1_test.go
Auto-merging pilot/pkg/security/authz/model/util_test.go
Auto-merging pilot/pkg/security/authz/model/util.go
Auto-merging pilot/pkg/security/authz/model/model.go
CONFLICT (content): Merge conflict in pilot/pkg/security/authz/model/model.go
Auto-merging pilot/pkg/security/authz/builder/integration_test.go
CONFLICT (content): Merge conflict in pilot/pkg/security/authz/builder/integration_test.go
Auto-merging pilot/pkg/security/authz/builder/builder_test.go
Patch failed at 0002 Add tests

```

hzxuzhonghu commented:
@pitlv2109 In windows file name can not contain *, so it failed when i pull the master recently.

pitlv2109 commented:
> @pitlv2109 In windows file name can not contain *, so it failed when i pull the master recently.

I'll change it. Thanks!
